ready for review (#3)
simplifiedcourses/simplifiedcourses.github.io: _posts/2022-12-19-running-outputs-outside-zonejs-for-angular-performance-optimization.md

---
layout: post
title:  "Running Outputs outside zone.js for Angular performance Optimization"
date:   2022-12-19
published: true
comments: true
categories: Angular, RxJS
cover: assets/running-outputs-outside-zonejs.png

---

In a [previous article](https://blog.simplified.courses/angular-change-detection-onpush-or-not/){:target="_blank"}, we covered the impact of the `OnPush` Change Detection strategy in Angular. We learned that when smartly applying this strategy, we can optimize our performance drastically.

It is also true that this optimization is not always the most important one we can take.
There are 3 ways we can optimize Change Detection in Angular applications:

> - Applying the `OnPush` Change Detection strategy.

Reviewer: It would be clear to use a numbered list here

Let's take a `mousemove` event for instance:
`ChildComponent` is responsible for giving its parent the `x` and `y` coordinates of the mouse when the user moves through an `@Output()`.
It listens to the `mousemove` event so every time that event is emitted (which is a dozen times per second), `zone.js` would pick this up and trigger Change Detection globally by executing the `tick()` function.
We could fix that issue by manually subscribing to that event in the outer zone:

```typescript
// Run in the outer zone
ngZone.runOutsideAngular(() => {
    fromEvent(this.document, 'mousemove')
        .subscribe((e: MouseEvent) => {
            this.notifyParent.emit({x: e.x, y: e.y});
        });
});
```

While this would not trigger Change Detection. It makes the use of `@Output()`s ugly and more complex.
It's not reactive and it makes the cleanup for subscriptions harder.
The fact that an `@Output()` can be assigned to any kind of Observable is quite nice.
Take this snippet for example:

```typescript
@Output() public readonly move = fromEvent(this.document, 'mousemove').pipe(
    map((e: MouseEvent) => {
        return {
            x: e.x,
            y: e.y,
        };
    }),

>     // debounceTime(1000),

Reviewer: Code in comments neccessary here?

### Quick recap

At this moment we are triggering Change Detection on the root of our application **a dozen times per second**. If we would have a big application that has a lot of components this could be problematic and even cause the application to crash.

## Binding Angular @Output() in the outer zone

We don't want to manually subscribe and run things in the outer zone... We want to keep the **reactive flow**. For that reason, we are going to create a custom operator that runs the subscription of this `@Output()` in the outer zone, so that Change Detection will not be triggered anymore.
We will create an `outsideAngular()` operator that we can use like this:

```typescript
@Output() public readonly valChange = interval(1000)
    // Subscribes to the source observable in the outer zone
    .pipe(outsideAngular());

@Output() public readonly move = fromEvent(this.document, 'mousemove')
    .pipe(

>         outsideAngular(), // Subscribes to the source observable

Reviewer 2: 🔥 That's brilliant

Author: Thanks!

                          // in the outer zone
        map((e: MouseEvent) => {
            return {
                x: e.x,
                y: e.y,
            };
        })
    );
```

Creating a custom operator in RxJS is quite straightforward. It's a function
that returns a function that takes the source observable as a parameter.

```typescript
export const outsideAngular = <T>() => (source$: Observable<T>) => {
    // Todo: create yourNewObservable$
    return yourNewObservable$;
};
```

In our case, we want to subscribe on `source$` inside the **outer zone** of Angular and keep the results of that subscription in a subject that we can return. 
For that, we need to get access to an instance of `NgZone`... We can use the `inject()` function of Angular to get access to that instance.

```typescript

> export const outsideAngular = <T>() => (source$: Observable<T>) => {

Reviewer 2: Keep the naming of `NgZone`. Easier to rememeber. :-)

```suggestion
export const runOutsideAngular = <T>() => (source$: Observable<T>) => {
```

    });

    // Return the subject but be sure to unsubscribe from
    // the source observable when the subject completes
    // or errors
    return sub$$.pipe(
        finalize(() => {
            subscription.unsubscribe();
        })
    );
};
```

Everything works as expected now: `ChildComponent` does not trigger Change Detection anymore but `ParentComponent` is notified with the new values.

We can see that in this [StackBlitz example](https://stackblitz.com/edit/angular-ivy-komr7r){:target="_blank"}
The DOM is not rerendered, but `console.log` statements are being made every second and when the user moves the mouse.

## Custom Change Detection

`ChildComponent` does what it needs to do:
- Run event listeners in the outer zone
- Notify the parent through `@Output()`s without triggering Change Detection

`ParentComponent` is now responsible of triggering Change Detection and we don't want to trigger it all the time. We only want to trigger it when the mouse of the user moves within the green box.

```typescript
export class ParentComponent {
    // Inject ChangeDetectorRef because we want to
    // manually trigger Change Detection with the
    // detectChanges() function

>     private readonly cdRef= inject(ChangeDetectorRef)

Reviewer 2: ```suggestion
    private readonly chnageDetectorRef= inject(ChangeDetectorRef)
```

Reviewer 2: Motivate your readers to not use abbreviations.

I personally go with `changeDetector`

Author: done

>     val = 0;

Reviewer: Public/private for consistency

Reviewer 2: +1 for consistency

I prefer not using `public` at all.

But this is a personal preference of mine. Consistency is key.

    mouseX = null;
    mouseY = null;

>     public onValChange(v: number): void {

Reviewer 2: ```suggestion
    public onValueChange(value: number): void {
```

        // Set the value of val, but don't trigger
        // Change Detection

>         this.val = v;

Reviewer 2: ```suggestion
        this.value = value;
```

        console.log(v);
    }

>     public onMove(e: { x: number; y: number }): void {

Reviewer 2: ```suggestion
    public onMouseMove(mouseCoordinates: { x: number; y: number }): void {
```

Author: Keeping move because (mousemove) might be another output

        // Set the values always
        this.mouseX = e.x;
        this.mouseY = e.y;
        // Only trigger Change Detection when the event
        // occured inside the green box
        if(e.x <= 100 && e.y <= 100){
            this.cdRef.detectChanges();
        }
        console.log(e);
    }
}
```

We can find the optimized Stackblitz example [here](https://stackblitz.com/edit/angular-ivy-zkbwrw){:target="_blank"}

## Conclusion

We learned that there are 3 ways of optimizing for Change Detection.

> The first way is by [using the OnPush Change Detection strategy which is described in detail in this article](https://blog.simplified.courses/angular-change-detection-onpush-or-not/){:target="_blank"}. The second one is detaching a component completely from the Change Detector and the third one is to run code in an outer zone.

Reviewer 2: Make this a list as well for quicker scanning...

1. By [using the OnPush....] ...
2. Detaching a component ....
3. Run Code in an outer....

We saw how we could achieve this by running `ngZone` its `runOutsideAngular()` function but that was hard to combine with reactive `@Output()`s.

Using Observables on `@Output()`s  gives us the ability to create reactive flows and by creating an `outsideAngular()` operator we were able to create `@Output()`s that notify their parents but do not initiate Change Detection.

While that is probably not something we want to use for everything, it might be beneficial for events that trigger Change Detection too much.

> If you like the article, leave me a comment below and if you want to learn more about Change Detection, you can support me by buying this [Angular Change Detection book](https://www.simplified.courses/angular-change-detection-simplified-e-book){:target="_blank"}.

Reviewer 2: ```suggestion
If you like the article, leave me a comment below and if you want to learn more about Change Detection, you migh be interesteed in buying this [Angular Change Detection book](https://www.simplified.courses/angular-change-detection-simplified-e-book){:target="_blank"}. It will help you understanding and resolveing performance issues in your Angular Enterprise projects.
```
